### PR TITLE
Added benchmarks for `head`, `tail`, `get`, `update`, `prepend`, `append`, `groupBy` and `iterate` from `List` and `Vector`

### PR DESCRIPTION
**Pull request by paplorinc**
Compared it against `ArrayList`, `LinkedList`, `PCollections`, `Functional Java`, `Scala Mutable` and `Scala Persistent`

> List

```java
Ratios slang / <alternative_impl>
Target         Operation   Ratio                                                    10        100       1000 
ListBenchmark  Head        slang_persistent/fjava_persistent                     1.08x      1.03x      1.02x 
ListBenchmark  Head        slang_persistent/java_mutable                         1.03x      1.04x      1.03x 
ListBenchmark  Head        slang_persistent/java_mutable_linked                  1.01x      0.99x      1.00x 
ListBenchmark  Head        slang_persistent/pcollections_persistent              1.12x      1.11x      1.12x 
ListBenchmark  Head        slang_persistent/scala_mutable                        1.04x      1.02x      1.02x 
ListBenchmark  Head        slang_persistent/scala_persistent                     1.00x      1.01x      1.00x 

ListBenchmark  Tail        slang_persistent/fjava_persistent                     0.98x      1.00x      0.97x 
ListBenchmark  Tail        slang_persistent/java_mutable                        24.40x     30.00x    148.05x 
ListBenchmark  Tail        slang_persistent/java_mutable_linked                  8.34x      9.68x     53.63x 
ListBenchmark  Tail        slang_persistent/pcollections_persistent              1.26x      1.30x      1.28x 
ListBenchmark  Tail        slang_persistent/scala_mutable                        5.44x      7.76x     50.02x 
ListBenchmark  Tail        slang_persistent/scala_persistent                     0.97x      1.01x      0.99x 

ListBenchmark  Append      slang_persistent/fjava_persistent                     0.39x      0.62x      0.67x 
ListBenchmark  Append      slang_persistent/java_mutable                         0.06x      0.01x      0.00x 
ListBenchmark  Append      slang_persistent/java_mutable_linked                  0.06x      0.01x      0.00x 
ListBenchmark  Append      slang_persistent/pcollections_persistent              0.33x      0.57x      0.98x 
ListBenchmark  Append      slang_persistent/scala_mutable                        0.11x      0.02x      0.00x 
ListBenchmark  Append      slang_persistent/scala_persistent                     0.25x      0.05x      0.00x 

ListBenchmark  Prepend     slang_persistent/fjava_persistent                     1.52x      1.65x      1.40x 
ListBenchmark  Prepend     slang_persistent/java_mutable                         4.62x      6.16x     16.36x 
ListBenchmark  Prepend     slang_persistent/java_mutable_linked                  1.04x      1.12x      1.11x 
ListBenchmark  Prepend     slang_persistent/pcollections_persistent              0.80x      0.86x      1.12x 
ListBenchmark  Prepend     slang_persistent/scala_mutable                        1.07x      1.07x      1.21x 
ListBenchmark  Prepend     slang_persistent/scala_persistent                     0.89x      1.04x      1.14x 

ListBenchmark  Get         slang_persistent/fjava_persistent                    18.65x      6.72x      4.79x 
ListBenchmark  Get         slang_persistent/java_mutable                         0.11x      0.00x      0.00x 
ListBenchmark  Get         slang_persistent/java_mutable_linked                  0.45x      0.22x      0.34x 
ListBenchmark  Get         slang_persistent/pcollections_persistent              1.36x      1.32x      1.61x 
ListBenchmark  Get         slang_persistent/scala_mutable                        1.12x      1.12x      1.03x 
ListBenchmark  Get         slang_persistent/scala_persistent                     0.77x      0.97x      0.99x 

ListBenchmark  Update      slang_persistent/java_mutable                         0.05x      0.00x      0.00x 
ListBenchmark  Update      slang_persistent/java_mutable_linked                  0.07x      0.03x      0.06x 
ListBenchmark  Update      slang_persistent/pcollections_persistent              0.50x      0.60x      0.90x 
ListBenchmark  Update      slang_persistent/scala_mutable                        0.13x      0.16x      0.17x

ListBenchmark  Iterate     slang_persistent/fjava_persistent                     1.09x      1.08x      0.98x 
ListBenchmark  Iterate     slang_persistent/java_mutable                         0.95x      0.46x      0.26x 
ListBenchmark  Iterate     slang_persistent/java_mutable_linked                  0.97x      0.83x      0.88x 
ListBenchmark  Iterate     slang_persistent/pcollections_persistent              0.84x      0.99x      0.98x 
ListBenchmark  Iterate     slang_persistent/scala_mutable                        1.41x      1.19x      0.61x 
ListBenchmark  Iterate     slang_persistent/scala_persistent                     0.94x      1.04x      1.00x 

ListBenchmark  GroupBy     slang_persistent/fjava_persistent                    10.63x     71.37x    226.17x 
ListBenchmark  GroupBy     slang_persistent/java_mutable                         0.41x      0.49x      0.73x 
ListBenchmark  GroupBy     slang_persistent/scala_persistent                     1.53x      1.25x      1.09x 
```

In conclusion: `List` is pretty fast, though it could use some minor optimizations here and there (e.g. `append` and `update` are still quite slow).

> Vector 

```java
Ratios slang / <alternative_impl>
Target           Operation  Ratio                                           10      100     1000 
VectorBenchmark  Head       slang_persistent/fjava_persistent            0.28x    0.27x    0.25x
VectorBenchmark  Head       slang_persistent/java_mutable                0.24x    0.22x    0.21x
VectorBenchmark  Head       slang_persistent/pcollections_persistent     0.65x    1.02x    1.50x
VectorBenchmark  Head       slang_persistent/scala_persistent            0.28x    0.34x    0.32x

VectorBenchmark  Tail       slang_persistent/fjava_persistent            7.77x   14.43x   10.47x
VectorBenchmark  Tail       slang_persistent/java_mutable                0.82x    0.84x    1.23x
VectorBenchmark  Tail       slang_persistent/pcollections_persistent     1.02x    1.65x    1.77x
VectorBenchmark  Tail       slang_persistent/scala_persistent            0.62x    1.01x    0.99x

VectorBenchmark  Get        slang_persistent/fjava_persistent            9.60x    7.73x   17.68x
VectorBenchmark  Get        slang_persistent/java_mutable                0.14x    0.03x    0.03x
VectorBenchmark  Get        slang_persistent/pcollections_persistent     1.39x    1.00x    2.54x
VectorBenchmark  Get        slang_persistent/scala_persistent            0.27x    0.14x    0.15x

VectorBenchmark  Update     slang_persistent/fjava_persistent           50.71x  163.88x  195.51x
VectorBenchmark  Update     slang_persistent/java_mutable                0.11x    0.05x    0.03x
VectorBenchmark  Update     slang_persistent/pcollections_persistent     1.22x    1.66x    2.08x
VectorBenchmark  Update     slang_persistent/scala_persistent            0.89x    0.62x    0.40x

VectorBenchmark  Prepend    slang_persistent/fjava_persistent            2.25x    2.77x    2.72x
VectorBenchmark  Prepend    slang_persistent/java_mutable                0.53x    0.55x    1.02x
VectorBenchmark  Prepend    slang_persistent/pcollections_persistent     0.97x    2.11x    2.88x
VectorBenchmark  Prepend    slang_persistent/scala_persistent            0.52x    0.48x    0.37x

VectorBenchmark  Append     slang_persistent/fjava_persistent            3.15x    3.44x    3.14x
VectorBenchmark  Append     slang_persistent/java_mutable                0.15x    0.10x    0.09x
VectorBenchmark  Append     slang_persistent/pcollections_persistent     1.33x    2.61x    4.20x
VectorBenchmark  Append     slang_persistent/scala_persistent            0.64x    0.51x    0.42x

VectorBenchmark  GroupBy    slang_persistent/java_mutable                0.21x    0.19x    0.25x
VectorBenchmark  GroupBy    slang_persistent/scala_persistent            0.89x    0.44x    0.30x

VectorBenchmark  Slice      slang_persistent/java_mutable                0.07x    0.00x    0.00x
VectorBenchmark  Slice      slang_persistent/scala_persistent            0.51x    0.07x    0.01x

VectorBenchmark  Iterate    slang_persistent/fjava_persistent           31.00x   55.51x   50.76x
VectorBenchmark  Iterate    slang_persistent/java_mutable                0.17x    0.07x    0.07x
VectorBenchmark  Iterate    slang_persistent/pcollections_persistent     1.95x    1.72x    1.51x
VectorBenchmark  Iterate    slang_persistent/scala_persistent            0.30x    0.19x    0.16x
```

### CODE REVIEW

File: javaslang-benchmark/src/test/java/javaslang/benchmark/JmhRunner.java [hunk 1]
```diff
@@ -9,38 +9,34 @@
...
+    private static final int DURATION_MILLIS = 500;
 
-    private static final int QUICK_WARMUP_ITERATIONS = 10;
-    private static final int QUICK_MEASUREMENT_ITERATIONS = 10;
-    public static final int QUICK_DURATION_MILLIS = 200;
-
-    public static void run(Class<?> benchmarkClass) {
-        runAndReport(benchmarkClass, WARMUP_ITERATIONS, MEASUREMENT_ITERATIONS, DURATION_MILLIS, PrintGc.Enable, Assertions.Disable, FORKS);
-    }
+    private static final int QUICK_WARMUP_ITERATIONS = 5;
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/JmhRunner.java]: for development purposes these are "good enough", compared to their speed

File: javaslang-benchmark/src/test/java/javaslang/benchmark/JmhRunner.java [hunk 2]
```diff
@@ -64,17 +60,6 @@ private static void runAndReport(Class<?> benchmarkClass, int warmupIterations,
         }
     }
 
-    private enum Assertions {
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/JmhRunner.java]: we don't want to measure the speed of assertions

File: javaslang-benchmark/src/test/java/javaslang/benchmark/collection/ArrayBenchmark.java [hunk 3]
```diff
@@ -87,39 +85,59 @@ public Object java_mutable(Initialized state) {
...
+        public Object slang_persistent() { return slangPersistent.tail(); }
     }
 
     public static class Get extends Base {
-        final int index = CONTAINER_SIZE / 2;
-
         @Benchmark
-        public Object java_mutable() { return javaMutable.get(index); }
+        public void java_mutable() {
+            for (int i = 0; i < ELEMENTS.length; i++) {
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/collection/ArrayBenchmark.java]: it seems more accurate to measure all the `get`s for all possible values, as the compiler seems to inline the value otherwise (for linked lists at least)

File: javaslang/src/main/java/javaslang/collection/List.java [hunk 4]
```diff
@@ -557,13 +557,19 @@
 
     @Override
     default List<T> append(T element) {
-        return foldRight(of(element), (x, xs) -> xs.prepend(x));
+        final ArrayList<T> results = new ArrayList<>(size() + 1);
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/List.java]: optimized slightly

File: javaslang/src/main/java/javaslang/collection/List.java [hunk 5]
```diff
@@ -668,18 +674,12 @@
...
-            throw new IndexOutOfBoundsException("get(" + index + ") on Nil");
-        }
-        if (index < 0) {
+        if (index < 0 || index >= length()) {
             throw new IndexOutOfBoundsException("get(" + index + ")");
         }
         List<T> list = this;
         for (int i = index - 1; i >= 0; i--) {
             list = list.tail();
-            if (list.isEmpty()) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/List.java]: no need to check this in every iteration as we're storing the sizes now :)

File: javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkPerformanceReporter.java [hunk 6]
```diff
@@ -5,55 +5,51 @@
...
+    private final Class<?> benchmarkClass;
     private final Collection<RunResult> runResults;
     private final String targetImplementation;
     private final double outlierLowPct;
     private final double outlierHighPct;
 
-    public static BenchmarkPerformanceReporter of(Collection<RunResult> runResults) {
-        return of(runResults, "slang", 0.3, 0.1);
+    public static BenchmarkPerformanceReporter of(Class<?> benchmarkClass, Collection<RunResult> runResults) {
+        return of(benchmarkClass, runResults, "slang", 0.3, 0.1);
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkPerformanceReporter.java]: It would be best if the display order would be the same as the order the groups were defined in, not alphabetically

File: javaslang/src/main/java/javaslang/collection/List.java (hunk 4 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/List.java]: Please do the List changes in a separate PR. This PR should contain only benchmark changes.

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/List.java]: Sure, though I need the benchmarks to be merged first, otherwise I don't know what to optimize.
I keep adding commits to this, because they depend on others that aren't merged yet :).

**Comment by danieldietrich** [comment]
Great, thank you! I have to apologize that I'm only able to react online in the evening hours...
